kernel/comps/virtio/src/transport/pci/legacy.rs

```diff
@@ -0,0 +1,318 @@
+const ISR_STATUS_OFFSET: usize = 0x13;
+// If MSI-X is enabled for the device, there are two additional fields.
+const CONFIG_MSIX_VECTOR_OFFSET: usize = 0x14;
+const QUEUE_MSIX_VECTOR_OFFSET: usize = 0x16;
+
+const DEVICE_CONFIG_OFFSET: usize = 0x14;
+const DEVICE_CONFIG_OFFSET_WITH_MSIX: usize = 0x18;
+
+pub struct VirtioPciLegacyTransport {
+    device_type: VirtioDeviceType,
+    common_device: PciCommonDevice,
+    config_bar: Bar,
+    num_queues: u16,
+    msix_manager: VirtioMsixManager,
+}
+
+impl VirtioPciLegacyTransport {
+    pub const QUEUE_ALIGN_SIZE: usize = 4096;
+
+    #[allow(clippy::result_large_err)]
+    pub(super) fn new(
+        common_device: PciCommonDevice,
+    ) -> Result<Self, (BusProbeError, PciCommonDevice)> {
+        let device_type = match common_device.device_id().device_id {
+            0x1000 => VirtioDeviceType::Network,
+            0x1001 => VirtioDeviceType::Block,
+            0x1002 => VirtioDeviceType::TraditionalMemoryBalloon,
+            0x1003 => VirtioDeviceType::Console,
+            0x1004 => VirtioDeviceType::ScsiHost,
+            0x1005 => VirtioDeviceType::Entropy,
+            0x1009 => VirtioDeviceType::Transport9P,
+            _ => {
+                warn!(
+                    "Unrecognized virtio-pci device id:{:x?}",
+                    common_device.device_id().device_id
+                );
+                return Err((BusProbeError::ConfigurationSpaceError, common_device));
+            }
+        };
+        info!("[Virtio]: Found device:{:?}", device_type);
+
+        let config_bar = common_device.bar_manager().bar(0).unwrap();
+
+        let mut num_queues = 0u16;
+        while num_queues < u16::MAX {
+            config_bar
+                .write_once(QUEUE_SELECT_OFFSET, num_queues)
+                .unwrap();
+            let queue_size = config_bar.read_once::<u16>(QUEUE_SIZE_OFFSET).unwrap();
+            if queue_size == 0 {
+                break;
+            }
+            num_queues += 1;
+        }
+
+        // TODO: Support interrupt without MSI-X
+        let mut msix = None;
+        for cap in common_device.capabilities().iter() {
+            match cap.capability_data() {
+                CapabilityData::Msix(data) => {
+                    msix = Some(data.clone());
+                }
+                _ => continue,
+            }
+        }
+        let Some(msix) = msix else {
+            return Err((BusProbeError::ConfigurationSpaceError, common_device));
+        };
+        let msix_manager = VirtioMsixManager::new(msix);
+
+        Ok(Self {
+            device_type,
+            common_device,
+            config_bar,
+            num_queues,
+            msix_manager,
+        })
+    }
+
+    pub(crate) fn calc_virtqueue_size_aligned(queue_size: usize) -> usize {
+        let align_mask = Self::QUEUE_ALIGN_SIZE - 1;
+
+        ((size_of::<Descriptor>() * queue_size + size_of::<u16>() * (3 + queue_size) + align_mask)
```

Comment:
Could you add some comments about the calculation? That will be helpful for future maintenance, thanks!